src/internal/message-port.ts

```diff
@@ -0,0 +1,52 @@
+let promise: Promise<MessagePort>
+
+/**
+ * Returns a MessagePort for one-on-one communication
+ *
+ * Depending on which context's code runs first, either an incoming port from the other side
+ * is accepted OR a port will be offered, which the other side will then accept.
+ */
+export const getMessagePort = (
+  thisContext: 'window' | 'content-script',
+  namespace: string,
+  onMessage: (e: MessageEvent<any>) => void,
+): Promise<MessagePort> => (
+  promise ??= new Promise((resolve) => {
+    const acceptMessagingPort = (event: MessageEvent) => {
+      const { data: { cmd, scope, context }, ports } = event
+      if (cmd === 'webext-port-offer' && scope === namespace && context !== thisContext) {
+        window.removeEventListener('message', acceptMessagingPort)
+        ports[0].onmessage = onMessage
+        ports[0].postMessage('port-accepted')
+        return resolve(ports[0])
+      }
+    }
+
+    const offerMessagingPort = () => {
+      const channel = new MessageChannel()
+      channel.port1.onmessage = (event: MessageEvent) => {
+        if (event.data === 'port-accepted') {
+          window.removeEventListener('message', acceptMessagingPort)
+          return resolve(channel.port1)
+        }
+
+        onMessage?.(event)
+      }
+
+      window.postMessage({
+        cmd: 'webext-port-offer',
+        scope: namespace,
+        context: thisContext,
+      }, '*', [channel.port2])
+    }
+
+    window.addEventListener('message', acceptMessagingPort)
+
+    // one of the contexts needs to be offset by at least 1 tick to prevent a race condition
+    // where both of them are offering, and then also accepting the port at the same time
+    if (thisContext === 'window')
+      setTimeout(offerMessagingPort, 0)
+    else
+      offerMessagingPort()
```

Comment:
doesn't this rely on the content script and the "window script" being initialized roughly at the same time? it seems like a potential constraint and not an obvious thing for the users - but maybe I'm misinterpreting how this is supposed to work.

After a really painful experience when developing a devtools panel I'm somewhat hesitant to trust the exact timing of those two scripts executed even if the user wants them to be executed immediately.

Reply:
No, in 99% of the cases content script will always be the first to load, it'll then inject a script tag into the page which will then become the window context. So in those 99% of the cases, content script will just listen for the message event with ports packaged with it from window, whenever it's ready.

This `setTimeout` ensures that even in 0.00001% chance that both content script and window code run at the same time, one of them is behind the other, at least in case of port exchange.

Reply:
> This setTimeout ensures that even in 0.00001% chance that both content script and window code run at the same time, one of them is behind the other, at least in case of port exchange.

I'm not worried about the cases in which those two execute at the same time but rather about cases when they start executing at completely different times. The raw API of the manifest/content scripts/etc doesn't really mandate the timing anyhow - one can freely execute any piece of code in any of those lazily/dynamically.

Reply:
> The raw API of the manifest/content scripts/etc doesn't really mandate the timing anyhow - one can freely execute any piece of code in any of those lazily/dynamically

Yes, and we accommodate for that. The content script that has `webext-bridge` code in it is free to execute before or after the injected window script that has `webext-bridge` in it, or vice versa. The execution of these contexts can be 3ms apart, or 1 hour apart.

When each of these context's code runs, they register the listener for an incoming port _from_ the other context, **AND** broadcast a port _for_ the other context. Example, if `window` context had already loaded, it'd mean it has already registered a listener for an incoming port, so whenever `content-script` loads, it'll also do a broadcast. Since `window` script has already been executed earlier, it'd pick the broadcasted port in the registered listener, and complete the docking. This process is same the other way around as well.